### DISCUSSION TITLE
doc packages

### DISCUSSION THREAD

**Discussion by mightyiam** [90-008/nix-cargo-integration#177] [Ideas]
I am finding myself in [some project](https://github.com/molybdenumsoftware/pr-tracker/) using crane to build `cargo doc` generated docs and export those as flake `packages`. I'm not sure anyone would ever build those docs. That project also publishes those docs:

https://molybdenumsoftware.github.io/pr-tracker/programs/pr_tracker_fetcher_config/struct.Environment.html
https://molybdenumsoftware.github.io/pr-tracker/programs/pr_tracker_api_config/struct.Environment.html

Perhaps nci could create a doc package per crate?

**Comment by 90-008** [comment]
we could do this yeah. having a `<project-name>-docs` output probably would be sufficient, `checkPhase` would run `cargo test --doc`.

**Reply by mightyiam** [reply]
A single package with all workspace docs by default sounds reasonable. Hrm... the `checkPhase` for that should be a no-op, I think. Because doctests are already a part of each crate's `checkPhase`, right?

**Reply by 90-008** [reply]
i forgot cargo test runs doc tests by default, don't need to include it then yeah

**Reply by mightyiam** [reply]
While that video is not published yet, we at [the Molybdenum Software show](https://www.youtube.com/@molybdenumsoftware) started trying to implement this. If @yusdacra you wish to quickly churn this out we wouldn't mind. Otherwise just informing you.

**Reply by 90-008** [reply]
i was working on it but it's quite a bit more complicated to do than i initially thought doing everything in the "right" way.

right now doing separate doc outputs for each crate would be pretty easy but having a cargo doc run over the entire project is complicated because of how dream2nix implements packaging. hooking out into crane directly is possible but i would rather not do that in NCI because it wouldn't be using dream2nix (there *is* a hacky solution i can potentially think of, but the artifacts wouldn't be cached properly with it which would cause more recompiles).

you can just call crane yourself as a workaround (the artifacts wouldn't be shared this way but you do get to run cargo doc on the whole project). or i just add separate cargo doc outputs and use something like https://lib.rs/crates/doc-merge to combine the doc outputs and add an index, which might also be fine?

**Reply by mightyiam** [reply]
I'd vote for "take your time and do it properly" :+1:

**Reply by 90-008** [reply]
well, the thing is, it could possibly take a long time since this could touch a lot of stuff in dream2nix potentially, and i probably won't be working on it in the near future as i don't consider it high prio enough to matter. i'll probably do the latter option and add doc packages for each crate and use doc-merge in a combined derivation, this way it'll "work as intended" and later on could replace it to be an actual cargo doc derivation.

**Reply by mightyiam** [reply]
At least in my case, I pass `--no-deps` and `--package foo --package bar` to list some specific packages.

**Reply by 90-008** [reply]
i pushed generating doc package for each crate, so it will be under the `docs` attribute of a crate's outputs. i'll probably look into how to combine those later, i have some ideas but for now this should be somewhat adequate. it does run with `--no-deps` since i can't think of any case where you'd want the deps really (other than convenience i suppose but if someone wants that they can just override the build phase /shrug)

**Reply by 90-008** [reply]
ok, just pushed combined docs. it will just put them under project outputs (named `docs`). can specify `docsIndexCrate` with a crate name that will have it's index.html be symlinked to the drv root to be more convenient to serve with some local http file server. it should hopefully work fine for more complex projects but i have not tested it with anything particularly complex.
